src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -848,6 +966,167 @@ NodeListInsertCommand(List *workerNodeList)
+		getObjectIdentityParts(address, &names, &args, false);
+		#else
+		objectType = getObjectTypeDescription(address);
+		getObjectIdentityParts(address, &names, &args);
+		#endif
+
+		if (!isFirstObject)
+		{
+			appendStringInfo(insertDistributedObjectsCommand, ", ");
+		}
+		isFirstObject = false;
+
+		appendStringInfo(insertDistributedObjectsCommand,
+						 "(%s, ARRAY[",
+						 quote_literal_cstr(objectType));
+
+		char *name = NULL;
+		bool firstInNameLoop = true;
+		foreach_ptr(name, names)
+		{
+			if (!firstInNameLoop)
+			{
+				appendStringInfo(insertDistributedObjectsCommand, ", ");
+			}
+			firstInNameLoop = false;
+			appendStringInfoString(insertDistributedObjectsCommand,
+								   quote_literal_cstr(name));
+		}
+
+		appendStringInfo(insertDistributedObjectsCommand, "]::text[], ARRAY[");
+
+		char *arg;
+		bool firstInArgLoop = true;
+		foreach_ptr(arg, args)
+		{
+			if (!firstInArgLoop)
+			{
+				appendStringInfo(insertDistributedObjectsCommand, ", ");
+			}
+			firstInArgLoop = false;
+			appendStringInfoString(insertDistributedObjectsCommand,
+								   quote_literal_cstr(arg));
+		}
+
+		appendStringInfo(insertDistributedObjectsCommand, "]::text[],");
+
+		if (distributionArgumentIndex == NULL)
+		{
+			appendStringInfo(insertDistributedObjectsCommand, "NULL, ");
+		}
+		else
+		{
+			appendStringInfo(insertDistributedObjectsCommand, "%d, ",
+							 *distributionArgumentIndex);
+		}
+		if (colocationId == NULL)
+		{
+			appendStringInfo(insertDistributedObjectsCommand, "NULL)");
+		}
+		else
+		{
+			appendStringInfo(insertDistributedObjectsCommand, "%d)",
+							 *colocationId);
+		}
+	}
+
+	appendStringInfo(insertDistributedObjectsCommand, ") ");
+
+	appendStringInfo(insertDistributedObjectsCommand,
+					 "SELECT citus_internal_add_object_metadata("
+					 "typetext, objnames, objargs, distargumentindex::int, colocationid::int) "
+					 "FROM distributed_object_data;");
+
+	return insertDistributedObjectsCommand->data;
+}
+
+
+/*
+ * citus_internal_add_object_metadata is an internal UDF to
+ * add a row to pg_dist_object.
+ */
+Datum
+citus_internal_add_object_metadata(PG_FUNCTION_ARGS)
+{
+	char *textType = TextDatumGetCString(PG_GETARG_DATUM(0));
+	ArrayType *nameArray = PG_GETARG_ARRAYTYPE_P(1);
+	ArrayType *argsArray = PG_GETARG_ARRAYTYPE_P(2);
+	int distributionArgumentIndexValue;
+	int colocationIdValue;
+	int *colocationId = NULL;
+	int *distributionArgumentIndex = NULL;
+
+	if (!PG_ARGISNULL(3))
+	{
+		distributionArgumentIndexValue = PG_GETARG_INT32(3);
+		distributionArgumentIndex = &distributionArgumentIndexValue;
+	}
+	if (!PG_ARGISNULL(4))
+	{
+		colocationIdValue = PG_GETARG_INT32(4);
+		colocationId = &colocationIdValue;
+	}
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* this UDF is not allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+	}
+
+	ObjectAddress objectAddress = PgGetObjectAddress(textType, nameArray,
```

Comment:
minor: would be nice to add a single line comment saying that acl/ownership checks are done inside.